sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -0,0 +1,362 @@
+            relation = new QueriedDocTable(table.tableRelation(), context.paths(), querySpec);
+            relation.normalize(context.analysisMetaData);
+        }
+        relation.analyzeWhereClause(context.analysisMetaData);
+        return relation;
+    }
+
+    @Override
+    public AnalyzedRelation visitMultiSourceSelect(MultiSourceSelect multiSourceSelect, Context context) {
+        MultiSourceSelect relation = multiSourceSelect;
+        if (context.querySpec != null) {
+            QuerySpec querySpec = mergeAndReplaceFields(multiSourceSelect, context.querySpec);
+            // must create a new MultiSourceSelect because paths and query spec changed
+            relation = new MultiSourceSelect(mapSourceRelations(multiSourceSelect), context.paths(), querySpec);
+        }
+        relation.pushDownQuerySpecs();
+        return relation;
+    }
+
+    private Map<QualifiedName, AnalyzedRelation> mapSourceRelations(MultiSourceSelect multiSourceSelect) {
+        return Maps.transformValues(multiSourceSelect.sources(), new com.google.common.base.Function<MultiSourceSelect.Source, AnalyzedRelation>() {
+            @Override
+            public AnalyzedRelation apply(MultiSourceSelect.Source input) {
+                return input.relation();
+            }
+        });
+    }
+
+    private QuerySpec mergeAndReplaceFields(QueriedRelation table, QuerySpec querySpec) {
+        QuerySpec mergedQuerySpec = mergeQuerySpec(querySpec, table.querySpec());
+        replaceFieldReferences(mergedQuerySpec);
+        return mergedQuerySpec;
+    }
+
+    private static QuerySpec mergeQuerySpec(@Nullable QuerySpec querySpec1, QuerySpec querySpec2) {
+        if (querySpec1 == null) {
+            return querySpec2;
+        }
+
+        QuerySpec querySpec = new QuerySpec()
+            .outputs(querySpec1.outputs())
+            .where(mergeWhere(querySpec1, querySpec2))
+            .orderBy(mergeOrderBy(querySpec1, querySpec2))
+            .offset(mergeOffset(querySpec1, querySpec2))
+            .limit(mergeLimit(querySpec1, querySpec2))
+            .groupBy(pushGroupBy(querySpec1, querySpec2))
+            .having(pushHaving(querySpec1, querySpec2))
+            .hasAggregates(querySpec1.hasAggregates() || querySpec2.hasAggregates());
+
+        return querySpec;
+    }
+
+    private static WhereClause mergeWhere(QuerySpec querySpec1, QuerySpec querySpec2) {
+        WhereClause where1 = querySpec1.where();
+        WhereClause where2 = querySpec2.where();
+
+        if (!where1.hasQuery() || where1 == WhereClause.MATCH_ALL) {
+            return where2;
+        } else if (!where2.hasQuery() || where2 == WhereClause.MATCH_ALL) {
+            return where1;
+        }
+
+        return new WhereClause(AndOperator.join(ImmutableList.of(where2.query(), where1.query())));
+    }
+
+    @Nullable
+    private static OrderBy mergeOrderBy(QuerySpec querySpec1, QuerySpec querySpec2) {
```

Comment:
function could be changed to take just `Optional<OrderBy> o1`, `Optional<OrderBy> o2` to make it easier to unit test